### ISSUE TITLE
Get docs up-to-date for bundles and resources for Hugo 0.35

### ISSUE THREAD

**Issue by bep** [gohugoio/hugoDocs#308]
A dugnad of sort.

We can use the next branch:

* https://github.com/gohugoio/hugoDocs/tree/next
* https://next--gohugoio.netlify.com/

@regisphilibert 
@kaushalmodi

I suggest a simple coordinated approach with focus on getting things done ...

**Comment by bep** [comment]
>I suggest a simple coordinated approach with focus on getting things done ...

As in: "I can do this and then you can do that ..."

And I much prefer short and precise documentation which is both easier to understand than the chatty variant. So don't hesitate to trim down on unneeded wording.

**Comment by regisphilibert** [comment]
@kaushalmodi we could split the work.

> And I much prefer short and precise documentation which is both easier to understand than the chatty variant. So don't hesitate to trim down on unneeded wording.

Don't worry I know this is not a blog post ;)

That being said, I'm working on updating the article with the new additions and I have a structure that I think work.

### I propose starting by the page bundle architecture (no cupcake involved)

### Then the available "properties" or info, not sure about the wording here 
- .Name 
- .Title
- .Permlink
- .RelPermalink
- .AbsSourceFilename (not sure about mentioning this one)

Of course we'll have to emphasize the importance of the .Name property regarding to Match/ByMatch and its potential overriding by metadata.

### Then the methods

- ByMatch (introducing glob)
- GetMatch
- ByType

Then we can have another big part on the metadata which are a big part of the feature and needs precise explaination (src param, order overwriting etc...

**Comment by bep** [comment]
>.AbsSourceFilename (not sure about mentioning this one)

I would leave it out; it is kind of an internal property.

**Comment by bep** [comment]
```
ByMatch (introducing glob)
GetMatch
ByType
```

The name is just ... `Match`. 

Also, I think image processing needs some attention.

**Comment by kaushalmodi** [comment]
How about making this whole Bundle section.. a **Bundle**?

Here's a broad-level content organization that I was thinking:

- Bundle
  - Page bundle (`index.md`)
  - Section (or Branch?) bundle (`_index.md`)
  - Resources
    - `resources` Params
    - Methods? Properties? (@bep What would be the Go-correct term for those?)
      - `.Name`, `.Title`, ..
    - Resource filtering methods
      - `.Match`, `.GetMatch`, ..
  - Image Processing

Documentation implementation detail: 

- The whole Bundle documentation would be actually a Bundle i.e. `content/../bundle/index.md`
- Page Bundle would be `content/../bundle/page-bundle`, Section Bundle would be `content/../bundle/section-bundle` (@bep or Branch Bundle?), and so on.

---

@regisphilibert 

> Then the available "properties" or info, not sure about the wording here
>  -  .Name
>  -  .Title

I think it would be useful to add `.ResourceType` to that list.

**Comment by regisphilibert** [comment]
I think it would be useful to add .ResourceType to that list.

Of couse, but we should mention that it's only the "main type" ?

**Comment by regisphilibert** [comment]
> How about making this whole Bundle section.. a Bundle?
Are you talking about sections or page bundle?

md file inside a bundle cannot have a standalone page, I had studied this use case.

We would need a special template for it? One that is able to loop through bundled md (called sections) and display their content in the template.

**Comment by kaushalmodi** [comment]
> md file inside a bundle cannot have a standalone page, I had studied this use case.

Correct. But that should be fine.. 

> We would need a special template for it? 

We can have a generic `single` layout for bundles that loops though all Resources pages, and set the `layout` front-matter in that bundle `index` front-matter. Turns out.. I was just working on that example.

**Comment by bep** [comment]
I suggest you make it a `section` ("Page Bundles"?), which is a bundle that can contain ... bundles. But don't make it too complex; the text is important.

**Comment by kaushalmodi** [comment]
> I suggest you make it a section

OK, how about something like this:

```
bundles/
├── image-processing
│   └── index.md
├── _index.md
├── page-bundle
│   └── index.md
├── resources
│   └── index.md
└── section-bundle
    └── index.md
```

I'd suggest calling the section just *Bundles* as in that, we will further distinguish between *Page* and *Section* bundles. WDYT?

**Comment by bep** [comment]
... but it would also be nice if "bundles and resources" wasn't separated away as "another thing". A bundle is a Page and a bundle can have many Resources.

**Comment by kaushalmodi** [comment]
The idea was to briefly explain page vs section bundles in their own pages.. `index` vs `_index`. Those will be short descriptions.

I kept `resources` separate as they would apply to both page and section bundles...and the same thing for image processing.

**Comment by kaushalmodi** [comment]
The `bundles/_index.md` can have a *summary content* that ties all those concepts together, with links to then individual `page-bundle`, `section-bundle`, `resources`, .. pages.

**Comment by regisphilibert** [comment]
@kaushalmodi I am a bit lost in you suggestion, but I trust you. 

Should we start a branch and both work on it?

Would you care to start adding the bundle/section to this branch and write which should be the first part of the doc : Bundle/File structure...

Then I could write the property/methods section

Then  we would treat metadata and Image processing (in this order)

@bep the image processing: Will it, at some point, be useable outside of page resources? Is it already ? And if so, shouldn't it have its own page in the doc.

**Comment by bep** [comment]
>Should we start a branch and both work on it?

Use the `next` branch.

**Comment by kaushalmodi** [comment]
@regisphilibert @bep OK, the rough skeleton is in: https://next--gohugoio.netlify.com/content-management/bundles/

**Comment by kaushalmodi** [comment]
Only one minor thing.. not sure why https://next--gohugoio.netlify.com/content-management/ is now not showing everything under content-management:

![image](https://user-images.githubusercontent.com/3578197/35350977-90c5d770-010d-11e8-8428-efc4a1961eba.png)

Compare that with https://gohugo.io/content-management/.

That said, I think this problem can be solved in parallel, and shouldn't block the Bundles documentation effort.

**Comment by regisphilibert** [comment]
ok I'm gonna start the "resources" part with methods, properties and metadata

**Comment by regisphilibert** [comment]
Anything of note as the difference between Name and Title other than the fact that Name is used by Match and GetMatch?
They're both by default the filename right? Or am I missing something?

**Comment by regisphilibert** [comment]
I don't want to derail the discussion but I can't figure out why Match never seems to catch .md files... Any idea ? I want to document this proper ;)

**Comment by regisphilibert** [comment]
Pushed my first draft on Page Resources: https://next--gohugoio.netlify.com/content-management/bundles/resources/

**Comment by regisphilibert** [comment]
@kaushalmodi moving on to Image processing, I'm not as knowledgeable in the matter so may need your help once first draft is out. Cheers.

**Comment by kaushalmodi** [comment]
@regisphilibert I am not knowledgeable in that area either, but am willing to help however I can. I need to yet be done with page/section bundle pages.. I see a lot of folks need to understand that concept.. saw few questions regarding that on discourse today itself.

**Comment by regisphilibert** [comment]
```
// Resize to a width of 600px and preserve ratio
{{ $image := $resource.Resize "600x" }} 

// Resize to a height of 400px and preserve ratio
{{ $image := $resource.Resize "x400" }} 
```
Just have to say I love this simple syntax where moving the x around just does what's expected. That's the kind of innovative human syntax that makes me love this framework a bit more everyday!
That may be some kind of well known syntax elsewhere, but not where I come from... :)

**Comment by regisphilibert** [comment]
Alright, Image Processing is on `next`. 
Most of it is taken from @bep's .32 intro from https://gohugo.io/about/new-in-032/
I edited reworded for better flow and expended explanation when thought useful. 
Cheers

**Comment by bep** [comment]
>That may be some kind of well known syntax elsewhere, but not where I come from... :)

It's an obvious syntax, but I don't think it's very common. Some people complained and suggested something in the line of "300h 200w" which is very hard on my eyes.

But having done my share of "legacy renaming" in a widely used software (Hugo) has made me think twice before releasing new API. Just a simple change from `RSSlink` to `RSSLink` cost a lot in support questions, even if we deprecated the old one in a controlled manner.

**Comment by regisphilibert** [comment]
It's perfect this way.

**Comment by bep** [comment]
The sunrise metadata example looks a little contrived and does not show this feature's potential.

I think a more practical variant of @kaushalmodi 's example in my commit here would be better:

https://github.com/gohugoio/hugo/commit/5a0819b9b5eb9e79826cfa0a65f235d9821b1ac4

**Comment by regisphilibert** [comment]
It's actually taken from my bogus/application use case from by blog [post](https://regisphilibert.com/blog/2018/01/hugo-page-resources-and-how-to-use-them/) that will soon be updated with this metadata exemple. 
It felt a bit pushy to shove my own material in the doc :/

**Comment by bep** [comment]
Reuse is always good. I guess it is clear enough for a starter. We can add some more examples later.

**Comment by regisphilibert** [comment]
So sunrise or application?

**Comment by regisphilibert** [comment]
On another matter, woudn't it make more sense to move "Page Resources doc" out of the content management section.
Bundles have their place there, but "resources" would better fit in the template section?
This is the way the Menu feature/package is documented

https://gohugo.io/templates/menu-templates/
https://gohugo.io/content-management/menus/

**Comment by kaushalmodi** [comment]
@regisphilibert 

> It's actually taken from my bogus/application use case from by blog post that will soon be updated with this metadata example.

Of course, I credited you [when I posted on discourse][1], and even on the [example post itself][2] 😄. The markdown source is [here][3] if you want to use as example.

> do you recommand the current sunrise or the bogus/application one?

He meant your application example for sure.

[1]: https://discourse.gohugo.io/t/resources-metadata-rule-ordering/10140/2?u=kaushalmodi
[2]: https://hugo-sandbox.netlify.com/posts/resources-order-a/
[3]: https://gitlab.com/kaushalmodi/hugo-sandbox/raw/8ba4c68bdc04e113995f8921aa013dd858aba5be/content/posts/resources-order-a/index.md

**Comment by kaushalmodi** [comment]
@bep It's possible that you missed [this comment earlier][1] among all the gohugoio organization messages that you would be getting. Can you please provide/suggest a good solution (details in that comment)?

[1]: https://github.com/gohugoio/hugoDocs/issues/308#issuecomment-360235461

**Comment by regisphilibert** [comment]
@kaushalmodi I was in no way implying you didn't credit :)
I will switch for you're improved application exemple but convert to .yaml, saves on lines.

**Comment by kaushalmodi** [comment]
> I will switch for you're improved application exemple but convert to .yaml

About that, I think it might be good to provide both YAML/TOML alternatives.. as the TOML version could look compicated (but is not).

**Comment by regisphilibert** [comment]
I don't know. It takes soooo much space on the page to include both. At some point, we should build a widget to allow tab toggling of the config languages. (so many place in the doc where both are added)

**Comment by kaushalmodi** [comment]
Something [like this][1] for the Hugo doc site would be nice, but that's a top of discussion for a later time. I have personally appreciated seeing TOML examples on the doc site as I prefer using TOML (and the same applies to YAML users). So it just seems fair to me to provide both examples especially for a complicated case like this one.

**Comment by bep** [comment]
>I don't know. It takes soooo much space on the page to include both. 

Toggling is out of scope for this issue. But for the `resources` definitions in front matter example would be good th have in both TOML and YAML, because that is actually hard to get right.

**Comment by bep** [comment]
>What's a good way to get a union of .Sections + .Paginator.Pages?

{{ .Sections | union .Paginator.Pages }}

**Comment by kaushalmodi** [comment]
Would that work for the rest of the doc site? Because this modification is needed in `_default/list.html`.

**Comment by bep** [comment]
>Would that work for the rest of the doc site? 

Do I look like the Oracle of Seasons? :-) I can look at it, eventually.

**Comment by kaushalmodi** [comment]
> Do I look like the Oracle of Seasons? :-)

Sorry :) I'll just commit it to `next` and we'll figure out that later.

**Comment by bep** [comment]
I see the use of "Page Bundles" and "Section Bundles". And while I admit that these need terms/names -- I think those are confusing.

* Everything is a Page.
* A Section is something special in Hugo
* A taxonomy list is a "bundle", but it's not a section

**Comment by regisphilibert** [comment]
Mmm... now I'm confused about the order of metadata rules. This is @bep exemple from his commit:
```toml
[[resources]]
src = "documents/photo_specs.pdf"
title = "Photo Specifications"
[resources.params]
ref = 90564687
icon = "photo"
[[resources]]
src = "documents/guide.pdf"
title = "Instruction Guide"
[resources.params]
ref = 90564568
[[resources]]
src = "documents/checklist.pdf"
title = "Document Checklist"
[resources.params]
ref = 90564572
[[resources]]
src = "documents/payment.docx"
title = "Proof of Payment"
[[resources]]
src = "documents/*.pdf"
title = "PDF file"
[resources.params]
icon = "pdf"
[[resources]]
src = "documents/*.docx"
title = "Word document"
[resources.params]
icon = "word"

```
The specific rule is first and yet is not overridden by the more generic rule following? Is it the correct behaviour?

**Comment by regisphilibert** [comment]
I updated the page accordingly. Ignore my previous comment, I re-read you commit message.

**Comment by kaushalmodi** [comment]
Hmm, how about 

- ~~Bundle~~ **Page** Bundle
  - ~~Page Bundle~~ **Leaf** Bundle
  - ~~Section Bundle~~ **Branch** Bundle

[[Reference][1]]

> A taxonomy list is a "bundle", but it's not a section

I didn't know of that thing.. what are *taxonomy lists*? Any relation with *Page Bundles*?

[1]: https://en.m.wikipedia.org/wiki/Tree_(data_structure)#Terminology_used_in_trees

**Comment by regisphilibert** [comment]
I keep thinking that the previous sunset exemple was more concise and readable. Disagree and commit ;)

**Comment by budparr** [comment]
Looks like you folks are demoing the page bundles with your docs on them. I personally think it's better to keep the docs consistent for future contributors and have demos elsewhere. At any rate, when I click around the next branch, and from this [comment](https://github.com/gohugoio/hugoDocs/issues/308#issuecomment-360235461) it looks like there might be some unintended consequences in the menu system, which was created back when bundles were merely a twinkle in BEP's eye. 

cc @rdwatters

**Comment by regisphilibert** [comment]
I tend to agree with @budparr here. Maybe revisit the doc as bundles later on.

**Comment by kaushalmodi** [comment]
Considering consistency, it makes sense to have the inner leaf bundles as regular pages instead, so that future contributors are not confused by the `index.md` files (hope they are not confused anyways). And also, those leaf bundles don't have to exist right now as there are no nested content pages or images. 

But I would like to retain the other branch bundle. That bundle helps organize all the Page Bundle related documentation together. 

Regarding the menu layout issue, it should be now fixed after my last commit. Even if that raises issues at other places in the site, it should be fixable. Eventually, we shouldn't prevent using these features because the layout needs to be tweaked a bit (or fixed within Hugo) to accommodate them. 

If we try to stick with the old layout, it would be unwieldy to have all those pages flat in the content-management Section. 

So, in summary, let's try to keep at least that outer Branch Bundle organization for the Page Bundles part.

**Comment by bep** [comment]
The "bundles as a new section" was not such a great idea. It 

1) Tries to establish bundles as "something special", which it isn't. Or, it should not be. It's just another way to organize your content.
2) It tries to create some new terms that does not exist, which I think will just confuse people
3) It does not work with the current menu system.

What I suggest is this:

* @regisphilibert if you could finish (that is an open term) the **content** of the Resources and Image Processing content page.
* Move those pages to the parent section `content-management`.

When that is done, I will try to write something that ties it all together in "Content Organization", which needs a trimming and overhaul anyway.

Sounds like a plan?

**Comment by kaushalmodi** [comment]
> Sounds like a plan?

OK. I have few comments. 

> It's just another way to organize your content.

If bundles is just a way to organize contents, why is it making the menu system complicated? Is here some fix needed in Hugo so that this can work in future (of course, this is an issue to tackle after the upcoming release). 

> It tries to create some new terms that does not exist

Wouldn't we need the new terms anyways? It will help better communicate with users. Something like "Look at Branch bundles section in the documentation" (because they used `index.md` instead of `_index.md`). 

I'll refactor the bundles documentation, everything into a single page, with leaf/branch terms, and then you can refactor that further as needed.

**Comment by regisphilibert** [comment]
@bep yes I will finish those. For me it means finalizing the metadata exemple with you guys suggestions. Let me know if you think I left out anything else. 

I’m still thinking the resources page belongs more in the template section than content management. Any thoughts on that?

**Comment by kaushalmodi** [comment]
> I’m still thinking the resources page belongs more in the template section than content management.

As this bundle structure is going away, I agree. Earlier it made sense as all the "page bundle" related stuff was in the same section.

Now I also think that Resources would better fit in the template section.

**Comment by regisphilibert** [comment]
> Now I also think that Resources would better fit in the template section.

Yes but my afterthought is the "Medatadata" part does belong in content management and I don't want to take it away from the resources page as both needs to be discuss alongside. Let's focus on finishing this for Monday. We can update later.

**Comment by regisphilibert** [comment]
So: b92959675c002d0cadfd606da53406be4b6e7c4d
and...
9978743ec4bf4205cc135510e0436ebde17dbc61

I updated the metadata example and renamed Resources to Page Resources. 
Now that this is out of the Bundles section, I think it makes more sense.

> When that is done, I will try to write something that ties it all together in "Content Organization", which needs a trimming and overhaul anyway.
Awesome

Today and this week-end won't give me as much time as this past week, but I will try to get an hour or two in the week end to follow up and help.

I appreciate being given the opportunity to help the Hugo community.

**Comment by kaushalmodi** [comment]
How does this look? https://next--gohugoio.netlify.com/content-management/page-bundles/

**Comment by kaushalmodi** [comment]
@regisphilibert I have made few edits to your Page Resources page. Can you review them: https://github.com/kaushalmodi/hugoDocs/commit/fb5b621dc63483babd8e27acf1768b122de705cd

In the process of ensuring the changes I made about the "order matters" part, I could have found an issue with using `:counter`.. it does not increment! @bep If you confirm this is a bug, I can open an issue on the hugo core repo.

### Test cases:

**TOML**
- [Markdown][1] | [Hugo Output][2]

**YAML**
- [Markdown][3] | [Hugo Output][4]

Note above, that even though I have `name : "pdf-file-:counter"` for `src : "**.pdf"`, the `name` for all PDF files stays stuck at `pdf-file-1`.. the counter does not increment.

[1]: https://gitlab.com/kaushalmodi/hugo-sandbox/raw/master/content/posts/resources-order-a/index.md
[2]: https://hugo-sandbox.netlify.com/posts/resources-order-a/#resource-debug
[3]: https://gitlab.com/kaushalmodi/hugo-sandbox/raw/master/content/posts/resources-order-a-yaml/index.md
[4]: https://hugo-sandbox.netlify.com/posts/resources-order-a-yaml/#resource-debug

**Comment by kaushalmodi** [comment]
OK.. as I experiment more, a weird pattern emerges..

- `title = "pdf-file-:counter"` ✅ --- Titles get set as expected: `pdf-file-1`, `pdf-file-2`, ..
- `name = "pdf-file-:counter"` 🚫 --- All names are set as `pdf-file-1`
- `foo = "foo-:counter"` (in `params`) 🚫 --- All `foo` get set as `foo-:counter` verbatim.. so no `:counter` replacement.

Shouldn't `:counter` way the same way as in `title` no matter in which param it is used.

**Comment by regisphilibert** [comment]
@kaushalmodi thanks as a non native speakers your proof reading and wording adjustments are of great help.

That being said, I think :counter is pretty self explanatory and your approach burdens this bullet whose purpose is just illustrate a metadata rule/src default setting on pdf files.

Maybe isolate :counter explanation in a notice afterward ?

Again, thanks for your corrections.

**Comment by kaushalmodi** [comment]
> is pretty self explanatory

I also assumed that, until https://github.com/gohugoio/hugoDocs/issues/308#issuecomment-360844191 happened :)

**Comment by regisphilibert** [comment]
> I also assumed that, until #308 (comment) happened :)

Let's wait on @bep's input on that before altering the doc on it.

**Comment by regisphilibert** [comment]
Beside if we’re not confident about :counter current behavior all the more reason to leave out of the doc for now.

**Comment by kaushalmodi** [comment]
@regisphilibert 

@bep fixed the `:counter` issue (https://github.com/gohugoio/hugo/issues/4335) today. I'll go ahead and merge my branch on the `next`, while moving the `:counter` description to a separate paragraph.

**Comment by kaushalmodi** [comment]
@regisphilibert 

> thanks as a non native speakers your proof reading and wording adjustments are of great help.

I couldn't have done that write-up better! I simply took care of my OCD, putting commas and dots :)

**Comment by kaushalmodi** [comment]
> merge my branch on the next by tomorrow

Done!

@bep I hit another snag while testing with a test case while coming up with the `:counter` documentation.. [details here][1].

[1]: https://github.com/gohugoio/hugo/issues/4335#issuecomment-361125556

**Comment by bep** [comment]
@kaushalmodi you need to **not** notify me about the same issue in duplicate copies/threads.

**Comment by kaushalmodi** [comment]
Fixed! 

- https://gohugo.io/content-management/page-bundles/
- https://gohugo.io/content-management/page-resources/
- https://gohugo.io/content-management/image-processing/
